File: services/PaperTrader.js

/**
 * Paper Trading Service
 * Wraps an exchange service to simulate execution while using real market data.
 */
export class PaperTrader {
    constructor(exchangeService, initialBalance = 100000) {
        this.exchange = exchangeService;
        this.portfolio = {
            cash: initialBalance,
            initialBalance,
            positions: {}, // ticker -> { quantity, avgPrice, entryTime }
            history: []
        };
        this.fees = {
            maker: 0.001, // 0.1%
            taker: 0.001
        };
    }

    // Pass-through for read-only data - uses getCandlesByTicker for proper ticker->symbolId resolution
    async getCandles(ticker, interval = '1m', startTime, endTime) {
        if (this.exchange.getCandlesByTicker) {
            return this.exchange.getCandlesByTicker(ticker, interval, startTime, endTime);
        }
        return this.exchange.getCandles(ticker, interval, startTime, endTime);
    }

    async getTicker(ticker) {
        // Try getQuoteByTicker first (Questrade)
        if (this.exchange.getQuoteByTicker) {
            try {
                const quote = await this.exchange.getQuoteByTicker(ticker);
                if (quote && quote.lastTradePrice) {
                    return { price: quote.lastTradePrice, bid: quote.bidPrice, ask: quote.askPrice };
                }
            } catch (e) { /* fall through */ }
        }

        if (this.exchange.getTicker) {
            return this.exchange.getTicker(ticker);
        }

        // Fallback: use latest candle
        try {
            const candles = await this.getCandles(ticker, '1m');
            if (candles && candles.length > 0) {
                const last = candles[candles.length - 1];
                return { price: last.c ?? last.close ?? 0 };
            }
        } catch (e) { /* fall through */ }

        return { price: 0 };
    }

    // Simulated Account Info
    async getBalance() {
        const equity = await this.calculateTotalEquity();
        return {
            cash: this.portfolio.cash,
            totalEquity: equity,
            marketValue: equity - this.portfolio.cash,
            buyingPower: this.portfolio.cash,
        };
    }

    async getPositions() {
        const positions = [];
        for (const [ticker, pos] of Object.entries(this.portfolio.positions)) {
            const priceData = await this.getTicker(ticker);
            const currentPrice = priceData.price || 0;
            const marketValue = pos.quantity * currentPrice;
            const openPnl = (currentPrice - pos.avgPrice) * pos.quantity;
            positions.push({
                symbol: ticker,
                openQuantity: pos.quantity,
                averageEntryPrice: pos.avgPrice,
                currentPrice,
                currentMarketValue: marketValue,
                openPnl,
                entryTime: pos.entryTime,
            });
        }
        return positions;
    }

    async getAccountSummary() {
        const balance = await this.getBalance();
        const positions = await this.getPositions();
        const pnl = balance.totalEquity - this.portfolio.initialBalance;
        const pnlPercent = (pnl / this.portfolio.initialBalance) * 100;
        return {
            ...balance,
            positions,
            pnl,
            pnlPercent,
            tradeCount: this.portfolio.history.length,
            initialBalance: this.portfolio.initialBalance,
        };
    }

    async calculateTotalEquity() {
        let equity = this.portfolio.cash;
        for (const [ticker, pos] of Object.entries(this.portfolio.positions)) {
            const priceData = await this.getTicker(ticker);
            const price = priceData.price || priceData.lastTradePrice || priceData.c || 0;
            equity += (pos.quantity * price);
        }
        return equity;
    }

    // Simulated Execution
    async createOrder(ticker, side, quantity, type = 'MARKET', price = null) {
        // 1. Get Real Price
        const priceData = await this.getTicker(ticker);
        const currentPrice = price || priceData.price || priceData.lastTradePrice || 0;

        if (!currentPrice) throw new Error(`Could not fetch price for ${ticker}`);

        const cost = currentPrice * quantity;
        const fee = cost * this.fees.taker;
        const totalCost = cost + fee;

        // 2. Validate
        if (side === 'BUY') {
            if (this.portfolio.cash < totalCost) {
                throw new Error(`Insufficient funds (Paper): Have $${this.portfolio.cash.toFixed(2)}, Need $${totalCost.toFixed(2)}`);
            }

            // Execute Buy
            this.portfolio.cash -= totalCost;

            if (!this.portfolio.positions[ticker]) {
                this.portfolio.positions[ticker] = { quantity: 0, avgPrice: 0, entryTime: Date.now() };
            }

            const pos = this.portfolio.positions[ticker];
            const newQty = pos.quantity + quantity;
            const newAvg = ((pos.quantity * pos.avgPrice) + cost) / newQty;

            pos.quantity = newQty;
            pos.avgPrice = newAvg;

        } else if (side === 'SELL') {
            const pos = this.portfolio.positions[ticker];
            if (!pos || pos.quantity < quantity) {
                throw new Error(`Insufficient position (Paper): Have ${pos ? pos.quantity : 0}, Need ${quantity}`);
            }

            // Execute Sell
            const revenue = (currentPrice * quantity) - fee;
            this.portfolio.cash += revenue;
            pos.quantity -= quantity;

            if (pos.quantity <= 0.00001) {
                delete this.portfolio.positions[ticker];
            }
        }

        // 3. Log
        const trade = {
            id: `paper-${Date.now()}`,
            ticker,
            side,
            quantity,
            price: currentPrice,
            fee,
            timestamp: Date.now()
        };
        this.portfolio.history.push(trade);

        console.log(`[PAPER TRADE] ${side} ${quantity} ${ticker} @ ${currentPrice} (Fee: ${fee.toFixed(2)})`);
        return trade;
    }

    // Reset paper trading state
    reset(initialBalance) {
        this.portfolio = {
            cash: initialBalance || this.portfolio.initialBalance,
            initialBalance: initialBalance || this.portfolio.initialBalance,
            positions: {},
            history: []
        };
    }

    getHistory() {
        return this.portfolio.history;
    }
}

File: canuck-trader-pro/backend/services/paper_trader.py
"""
Paper Trader for Questrade
Port of services/PaperTrader.js.

Virtual portfolio with cash + positions for paper trading stocks.
"""

import time
import logging

logger = logging.getLogger("paper_trader")


class PaperTrader:
    def __init__(self, questrade_service, starting_cash: float = 100000):
        self.questrade = questrade_service
        self.cash = starting_cash
        self.starting_cash = starting_cash
        self.positions: dict[str, dict] = {}  # ticker -> position
        self.trade_history: list[dict] = []

    def get_positions(self) -> list[dict]:
        result = []
        for ticker, pos in self.positions.items():
            result.append({
                "symbol": ticker,
                "quantity": pos["quantity"],
                "avg_entry": pos["avg_entry"],
                "current_value": pos["quantity"] * pos.get("last_price", pos["avg_entry"]),
                "unrealized_pnl": (pos.get("last_price", pos["avg_entry"]) - pos["avg_entry"]) * pos["quantity"],
            })
        return result

    def get_account_summary(self) -> dict:
        positions_value = sum(
            p["quantity"] * p.get("last_price", p["avg_entry"])
            for p in self.positions.values()
        )
        total = self.cash + positions_value
        pnl = total - self.starting_cash
        pnl_pct = (pnl / self.starting_cash * 100) if self.starting_cash > 0 else 0

        return {
            "cash": round(self.cash, 2),
            "positions_value": round(positions_value, 2),
            "total_value": round(total, 2),
            "starting_cash": self.starting_cash,
            "pnl": round(pnl, 2),
            "pnl_percent": round(pnl_pct, 2),
            "position_count": len(self.positions),
            "trade_count": len(self.trade_history),
        }

    def buy(self, ticker: str, quantity: float, price: float) -> dict:
        cost = quantity * price
        if cost > self.cash:
            return {"success": False, "error": "Insufficient funds"}

        self.cash -= cost

        if ticker in self.positions:
            pos = self.positions[ticker]
            total_qty = pos["quantity"] + quantity
            total_cost = pos["avg_entry"] * pos["quantity"] + cost
            pos["avg_entry"] = total_cost / total_qty
            pos["quantity"] = total_qty
            pos["last_price"] = price
        else:
            self.positions[ticker] = {
                "quantity": quantity, "avg_entry": price,
                "last_price": price, "open_time": time.time(),
            }

        trade = {
            "action": "BUY", "ticker": ticker, "quantity": quantity,
            "price": price, "cost": cost, "time": time.time(),
        }
        self.trade_history.append(trade)
        logger.info(f"Paper BUY: {quantity} {ticker} @ {price:.2f}")
        return {"success": True, "trade": trade}

    def sell(self, ticker: str, quantity: float, price: float) -> dict:
        pos = self.positions.get(ticker)
        if not pos or pos["quantity"] < quantity:
            return {"success": False, "error": f"Insufficient position in {ticker}"}

        revenue = quantity * price
        pnl = (price - pos["avg_entry"]) * quantity
        self.cash += revenue
        pos["quantity"] -= quantity
        pos["last_price"] = price

        if pos["quantity"] <= 0:
            del self.positions[ticker]

        trade = {
            "action": "SELL", "ticker": ticker, "quantity": quantity,
            "price": price, "revenue": revenue, "pnl": round(pnl, 2),
            "pnl_percent": round((price - pos["avg_entry"]) / pos["avg_entry"] * 100, 2),
            "time": time.time(),
        }
        self.trade_history.append(trade)
        logger.info(f"Paper SELL: {quantity} {ticker} @ {price:.2f} (PnL: {pnl:.2f})")
        return {"success": True, "trade": trade}

    def update_prices(self, price_map: dict[str, float]):
        for ticker, price in price_map.items():
            if ticker in self.positions:
                self.positions[ticker]["last_price"] = price

    def get_history(self, limit: int = 100) -> list[dict]:
        return self.trade_history[-limit:]

    def reset(self, starting_cash: float | None = None):
        self.cash = starting_cash or self.starting_cash
        self.starting_cash = self.cash
        self.positions.clear()
        self.trade_history.clear()
        logger.info(f"Paper portfolio reset to ${self.cash:.2f}")
        return {"success": True, "cash": self.cash}
